rest/diagnostic_api.go

```diff
@@ -0,0 +1,69 @@
+//  Copyright 2013-Present Couchbase, Inc.
+//
+//  Use of this software is governed by the Business Source License included
+//  in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+//  in that file, in accordance with the Business Source License, use of this
+//  software will be governed by the Apache License, Version 2.0, included in
+//  the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"fmt"
+
+	"github.com/couchbase/sync_gateway/auth"
+	"github.com/couchbase/sync_gateway/base"
+	channels "github.com/couchbase/sync_gateway/channels"
+	"github.com/gorilla/mux"
+)
+
+type allChannels struct {
+	Channels map[string]map[string]auth.GrantHistory `json:"all_channels,omitempty"`
+}
+
+func (h *handler) handleGetAllChannels() error {
+	h.assertAdminOnly()
+	user, err := h.db.Authenticator(h.ctx()).GetUser(internalUserName(mux.Vars(h.rq)["name"]))
+	if err != nil {
+		return fmt.Errorf("could not get user %s: %w", user.Name(), err)
+	}
+	if user == nil {
+		return kNotFoundError
+	}
+
+	resp := make(map[string]map[string]auth.GrantHistory)
+
+	// handles deleted collections, default/ single named collection
+	for _, dsName := range h.db.DataStoreNames() {
+		keyspace := dsName.ScopeName() + "." + dsName.CollectionName()
+
+		currentChannels := user.InheritedCollectionChannels(dsName.ScopeName(), dsName.CollectionName())
+		chanHistory := user.CollectionChannelHistory(dsName.ScopeName(), dsName.CollectionName())
+		// If no channels aside from public and no channels in history, don't make a key for this keyspace
+		if len(currentChannels) > 1 || len(chanHistory) != 0 {
+			resp[keyspace] = make(map[string]auth.GrantHistory)
+		}
```

Comment:
```suggestion
		// If no channels aside from public and no channels in history, this collection has no grant history. '!' will always be including in currentChannels
		if len(currentChannels) > 1 || len(chanHistory) != 0 {
			continue
		}
		resp[keyspace] = make(map[string]auth.GrantHistory)
```

I think an early exit makes this more clear.